include/seastar/core/abort_source.hh

```diff
@@ -112,8 +112,10 @@ public:
 
     /// Requests that the target operation be aborted. Current subscriptions
     /// are invoked inline with this call, and no new ones can be registered.
+    /// Must be called only once, otherwise terminate will be called.
     void request_abort() noexcept {
-        std::exchange(_subscriptions, std::nullopt)->clear_and_dispose([] (subscription* s) { s->on_abort(); });
+        auto subs = std::exchange(_subscriptions, std::nullopt).value();
+        subs.clear_and_dispose([] (subscription* s) { s->on_abort(); });
```

Comment:
The PR description says "throw" but this says "terminate".

Why not make the second request_abort() a no-op? Seems safest.

Reply:
I don't know about no-op. The caller expects abort to be requested but it was already requested and the subscribers no longer listen to the abort source, so a no-op would misleading.

Callers can easily avoid this if the request abort only if not requested already.

Reply:
@xemul / @gleb-cloudius / @nyh I don't mind much making request_abort noop if it's called more than once.
Any objections?

Reply:
Calling request_abort() from several places sounds like API mis-use, so I'd red-flag this early.

Reply:
Why? We already have cases like this, e.g. we abort either on timeout, or when operation completes. Kamil's testing framework does this.

Reply:
I do not think he calls abort when operation completes.  Anyway he calls it only once and as far as I understand @xemul meant that calling it several times and expect it to work is a misuse, not that the call should be in one place only.

Reply:
That code shows intent that the abort supposed to be called only once. If you hide the if inside you can do it without really intending it. I do not think anybody disagrees that having a call in more then one place is not allowed. Though I would use two different abort sources here instead of one for two different fibers.

Reply:
Is there an example of code which wouldn't intend that? abort_source is not used for message passing, it's an edge-triggered notification mechanism. The callback is intended to be called on the edge, at most once. 

You could use two abort sources and guard with ifs, but it makes the code more complex and error prone. I don't see why we should pay this price.

Reply:
I could imagine a service that have two ways to abort (on shutdown or on drain for instance) at they are slightly different. Each one triggers an abort but you do not want to run both of them.

With two abort source I do not see why would we need to guard then with ifs because each one will be triggered from one place.

Reply:
> I could imagine a service that have two ways to abort (on shutdown or on drain for instance) at they are slightly different. Each one triggers an abort but you do not want to run both of them.
> 

How is it related to abort_source::request_abort?

> With two abort source I do not see why would we need to guard then with ifs because each one will be triggered from one place.

True